### ISSUE TITLE
Stator halting, not processing items until a restart

### ISSUE THREAD

**Issue by rcarmo** [jointakahe/takahe#424]
I've been having a weird behavior where the stator will just lie there and do nothing for half a day or so, and then when I notice the timeline isn't updated and restart it it will churn through hundreds of items. 

There is nothing in the logs (no error message, nothing) and I don't see any difference in CPU usage from it (it's not in a loop). The only thing I can see  in the log is a "Running cleaning and scheduling" line with... nothing following it.

Maybe add some kind of watchdog timer is in order to avoid it getting stuck while processing something? If it exited after timeout, my `Procfile` handler (and docker-compose) would restart it automatically.

**Comment by manfre** [comment]
Stator shouldn't block. Most things are deferred async tasks that don't wait for completion. What's the commandline you're using to run stator? Does it block after the first "Running cleaning and scheduling" or at some random point after it's been running for a while? Does it seem to stop processing after a certain amount of time after starting (or no pattern)?

**Comment by rcarmo** [comment]
I am running the stator with this command line (I use [Piku](https://github.com/piku/piku) as a supervisor, and have a [writeup here](https://taoofmac.com/space/blog/2022/12/21/0900)).

```Procfile
wsgi: takahe.wsgi:application
worker: python manage.py runstator
release: python manage.py migrate
```

I also have `PYTHONUNBUFFERED=1` set, and when I looked at the process it was taking up less than 1% CPU. I didn't see any significant amount of outbound sockets to ports 80 or 443 in `netstat`, so I assume that if it was a network condition (which seems more likely) it might have been a DNS resolution issue (which can be tested for and logged against).

All the 2-3 times this has happened have been spotted in the morning, as I peek at the logs (I am running an instance off `git` tip, so I update it every day). I saw a lot of HTTP inbox requests, but nothing from the stator (which pops up every few minutes or so), so I investigated.

I pulled `git` tip, restarted everything (which in my case also includes running the migrations as a `release` worker), and the stator spent 5-10 minutes flushing out hundreds of pending events.

**Comment by andrewgodwin** [comment]
Yeah, I've seen this a couple of times with takahe.social too. We're pushing Django's async support to near its limit with the amount of sync-to-async-to-sync transitions we do in there daily, so I suspect there is some sort of eventual deadlock that happens.

There is a built-in liveness check for Stator that you can enable if your system supports command-based liveness checks, which is what we now use, but that's not going to be ideal for everyone.

I'll also go now and see if I can set up a SIGALRM watchdog that exits if it does get stuck. I think that's probably a reasonable approach.

**Comment by andrewgodwin** [comment]
Yeah, it might well be that too - the main thing is that it is incredibly hard to replicate and debug, and these kinds of issues can reappear at any time, so we need a watchdog in there anyway.

**Comment by rcarmo** [comment]
Maybe dump a `pdb` file when it stalls?  Haven't used that trick in a while.

**Comment by andrewgodwin** [comment]
So I just fixed a few nasty things in Stator I've found as takahe.social has been getting larger - it may well have been the source of the halting, but there is of course no way to tell without waiting a few weeks and seeing how many restarts happen.

**Comment by futzle** [comment]
Hi everyone, brand new user here. I'm running cockatoot.city on a Linode 1 GB Nanode, and I think I just hit this bug. I'm new to Docker so I'm still unsure how to get more than what `docker logs -t worker` tells me. New posts remained in the "new" state according to the Django admin interface, and never entered the "fanned_out" state.

Stopping and restarting the stator container seems to have unblocked things.

(Pre-emptively agree, 1 GB is pretty constrained. I'm trying out Tahakē as a low-resource alternative to Mastodon, which I also run.)

**Comment by andrewgodwin** [comment]
Yeah, this is still occasionally an issue for me as well - I'm actually partway through a rewrite of Stator to remove most of the async code, which is what I suspect is the root cause of all this. Hopefully when that's landed (another week or two I'd guess) you can give it a spin and see if it helps.

1GB is pretty constrained, I agree, but I'd hope we can get it to work reliably - nothing _should_ take more than that.
